core/message_queue.go

```diff
@@ -174,6 +183,30 @@ func (mq *MessageQueue) Size() int64 {
 	return l
 }
 
+// Oldest returns the oldest message stamp in the MessageQueue.
+// Oldest returns 0 if the queue is emptystarts
+func (mq *MessageQueue) Oldest() uint64 {
+	if mq.Size() == 0 {
+		return 0
+	}
+
+	mq.lk.Lock()
+	defer mq.lk.Unlock()
+
+	var all []*QueuedMessage
+	for _, qm := range mq.queues {
+		for _, m := range qm {
+			all = append(all, m)
+		}
+	}
+
+	sort.Slice(all, func(i, j int) bool {
+		return all[i].Stamp < all[j].Stamp
+	})
```

Comment:
This is way more work than necessary for finding the single smallest value. Just loop and maintain the least so far, perhaps in the named return value. The early exit for `Size() == 0` will then be more expensive than the normal case, I expect, so drop it.